src/main/java/edu/harvard/iq/dataverse/engine/command/impl/FinalizeDatasetPublicationCommand.java

```diff
@@ -399,16 +436,17 @@ private void notifyUsersFileDownload(CommandContext ctxt, DvObject subject) {
             .filter(  ra -> ra.getRole().permissions().contains(Permission.DownloadFile) )
             .flatMap( ra -> ctxt.roleAssignees().getExplicitUsers(ctxt.roleAssignees().getRoleAssignee(ra.getAssigneeIdentifier())).stream() )
             .distinct() // prevent double-send
-            .forEach( au -> ctxt.notifications().sendNotification(au, getTimestamp(), UserNotification.Type.GRANTFILEACCESS, getDataset().getId()) );
+            .forEach( au -> ctxt.notifications().sendNotificationInTransaction(au, getTimestamp(), UserNotification.Type.GRANTFILEACCESS, getDataset().getId()) );
     }
     
-    private void notifyUsersDatasetPublish(CommandContext ctxt, DvObject subject) {
+    private void notifyUsersDatasetPublishStatus(CommandContext ctxt, DvObject subject, UserNotification.Type type) {
```

Comment:
I thought I understood why the in transaction version of sendNotification was needed (to send the email even if the command transaction was rolling back), but this looks like the opposite - is it correct as is? (If so, what am I missing?)

Reply:
No, you were right. Since the notification needs to be saved in the database, rolling back the transaction un-saves it. And now that, in addition to the success notification, I am also sending notifications in situations when something goes bad and the transaction is going to be rolled back, I need to save the notification in a transaction of its own. 
Of course I should probably  only be using the "InTransaction" method when it is an on-failure notification; and stick with the normal, old way of doing it when it's the "success" notification at the end of the command. When nothing is expected to be rolled back that is. 

The extra parameter - type - identifies the type of the notification that needs to be sent;  "dataset's been published" vs "failed on account of registration error" ...

Reply:
(so yeah, I wasn't sure if you were familiar with how the notifications work - it's no just the email; the email that gets sent during the execution of the command doesn't get un-sent when the transaction is rolled back - :) - but we also save them in a database table, for the ui display)

Reply:
Right - but notifyUserDatasetPublishStatus is the one being used for the failure message, and it's using sendNotification inside and not sendNotificationInTransaction - that means the notification won't get saved in the db for the failure? Or?

Reply:
Did I check in the wrong version of it? - hold on...

Reply:
Of course I did. (I was messing with renaming the methods at the last moment - and did mess it up). Thank you.

Reply:
been there, done that!

Reply:
I was only looking at the line your comment was immediately referencing and got confused. 
OK, fixed - thanks again. 
Otherwise Kevin would run into this during QA, with no notification showing in the UI.